crates/dyn-abi/src/eip712/parser.rs

```diff
@@ -139,13 +141,61 @@ impl<'a> EncodeType<'a> {
 
         Ok(Self { types })
     }
+
+    /// Computes the canonical string representation of the type.
+    ///
+    /// Orders the `ComponentTypes` based on the EIP-712 rules, removes unsupported whitespaces, and
+    /// validates them.
+    pub fn canonicalize(&self) -> Result<String, Error> {
+        // Ensure no unintended whitespaces
+        let mut resolver = Resolver::default();
+        for component_type in &self.types {
+            resolver.ingest(component_type.to_owned());
+        }
+
+        // Resolve and validate non-dependent types
+        let mut non_dependent = resolver.non_dependent_types();
+
+        let first = non_dependent
+            .next()
+            .ok_or_else(|| Error::MissingType(String::from("primary component")))?;
```

Comment:
```suggestion
            .ok_or_else(|| Error::MissingType("primary component".to_string()))?;
```

might night `alloc::string::ToString` import